This project will leverage location data to find a ideal district to open a hotpot restaurant in HK.

This capstone project is designed to solve a question: which place is an ideal place to open a coffee shop in Toronto?

In [136]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

print('Libraries imported.')

Libraries imported.


In [137]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

In [138]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(website_url,'lxml')
print(soup.prettify())
My_table = soup.find('table',{'class':'wikitable sortable'})
columns = [th.text.replace('\n', '') for th in My_table.find('tr').find_all('th')]
trs = My_table.find_all('tr')[1:]
rows = list()
for tr in trs:
    rows.append([td.text.replace('\n', '').replace('\xa0', '') for td in tr.find_all('td')])
rows[:5]
df_Toronto = pd.DataFrame(data=rows, columns=columns)
indexNames = df_Toronto[df_Toronto['Borough'] == 'Not assigned' ].index
df_Toronto.drop(indexNames, inplace=True)
df_Toronto = df_Toronto.groupby('Postcode').agg({'Borough':'first', 'Neighbourhood': ', '.join}).reset_index()

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className=document.documentElement.className.replace(/(^|\s)client-nojs(\s|$)/,"$1client-js$2");RLCONF={"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":906439794,"wgRevisionId":906439794,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":!1,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June",

In [139]:
LaLo = pd.read_csv('http://cocl.us/Geospatial_data')
df_Toronto.columns = ['Postal Code','Borough','Neighbourhood']
df_Toronto_LoLa = pd.merge(df_Toronto,LaLo, on = 'Postal Code')
value_list = ['West Toronto', 'East Toronto', 'Downtown Toronto']
Toronto_data = df_Toronto_LoLa[df_Toronto_LoLa['Borough'].isin(value_list)].reset_index(drop=True)

In [140]:
CLIENT_ID = '0PCSZ3NLJLPZKYIIEEHLZ4QTP0DSKAOGFGQYCWYKRQQ3QDM0' # your Foursquare ID
CLIENT_SECRET = '1QW5XWTQADXVHMDTFVDUAPILVSTFBAELDIEWZJ0TL0S1CPPK' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)
LIMIT = 50
radius = 500

Your credentails:
CLIENT_ID: 0PCSZ3NLJLPZKYIIEEHLZ4QTP0DSKAOGFGQYCWYKRQQ3QDM0
CLIENT_SECRET:1QW5XWTQADXVHMDTFVDUAPILVSTFBAELDIEWZJ0TL0S1CPPK


In [141]:

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [142]:
Toronto_venues = getNearbyVenues(names=Toronto_data['Neighbourhood'],latitudes=Toronto_data['Latitude'],longitudes=Toronto_data['Longitude'])


The Beaches
The Danforth West, Riverdale
The Beaches West, India Bazaar
Studio District
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The Junction South
Parkdale, Roncesvalles
Runnymede, Swansea
Business Reply Mail Processing Centre 969 Eastern


In [143]:
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  Airport  Airport Food Court  Airport Gate  \
0                   The Beaches        0                   0             0   
1                   The Beaches        0                   0             0   
2                   The Beaches        0                   0             0   
3                   The Beaches        0                   0             0   
4  The Danforth West, Riverdale        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  Beer Bar  \
0       0     0    0                 0                   0      0         0   
1       0     0    0                 0                   0      0         0   
2       0     0    0                 0                   0      0         0   
3       0     0    0                 0                   0      0         0   
4       0     0    0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bistro  Board Shop  Boat or Ferry  \
0           0                   0       0           0              0   
1           0                   0       0           0              0   
2           0                   0       0           0              0   
3           0                   0       0           0              0   
4           0                   0       0           0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Butcher  Café  \
0                0         0             0              0        0     0   
1                0         0             0              0        0     0   
2                0         0             0              0        0     0   
3                0         0             0              0        0     0   
4                0         0             0              0        0     0   

   Cajun / Creole Restaurant  Caribbean Restaurant  Cheese Shop  \
0                          0                     0            0   
1                          0                     0            0   
2                          0                     0            0   
3                          0                     0            0   
4                    

In [144]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped

Neighbourhood  Airport  \
0                            Adelaide, King, Richmond   0.0000   
1                                         Berczy Park   0.0000   
2        Brockton, Exhibition Place, Parkdale Village   0.0000   
3   Business Reply Mail Processing Centre 969 Eastern   0.0000   
4   CN Tower, Bathurst Quay, Island airport, Harbo...   0.0625   
5                         Cabbagetown, St. James Town   0.0000   
6                                  Central Bay Street   0.0000   
7           Chinatown, Grange Park, Kensington Market   0.0000   
8                                            Christie   0.0000   
9                                Church and Wellesley   0.0000   
10                     Commerce Court, Victoria Hotel   0.0000   
11           Design Exchange, Toronto Dominion Centre   0.0000   
12                       Dovercourt Village, Dufferin   0.0000   
13             First Canadian Place, Underground city   0.0000   
14                     Harbord, University of Toronto   0.0000   
15  Harbourfront East, Toronto Islands, Union Station   0.0000   
16                          Harbourfront, Regent Park   0.0000   
17                      High Park, The Junction South   0.0000   
18                           Little Portugal, Trinity   0.0000   
19                             Parkdale, Roncesvalles   0.0000   
20                                           Rosedale   0.0000   
21                                 Runnymede, Swansea   0.0000   
22                           Ryerson, Garden District   0.0000   
23                                     St. James Town   0.0000   
24                    Stn A PO Boxes 25 The Esplanade   0.0000   
25                                    Studio District   0.0000   
26                                        The Beaches   0.0000   
27                     The Beaches West, India Bazaar   0.0000   
28                       The Danforth West, Riverdale   0.0000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0               0.0000        0.0000           0.000           0.0000   
1               0.0000        0.0000           0.000           0.0000   
2               0.0000        0.0000           0.000           0.0000   
3               0.0000        0.0000           0.000           0.0000   
4               0.0625        0.0625           0.125           0.1875   
5               0.0000        0.0000           0.000           0.0000   
6               0.0000        0.0000           0.000           0.0000   
7               0.0000        0.0000           0.000           0.0000   
8               0.0000        0.0000           0.000           0.0000   
9               0.0000        0.0000           0.000           0.0000   
10              0.0000        0.0000           0.000           0.0000   
11              0.0000        0.0000           0.000           0.0000   
12              0.0000        0.0000           0.000           0.0000   
13              0.0000        0.0000           0.000           0.0000   
14              0.0000        0.0000           0.000           0.0000   
15              0.0000        0.0000           0.000           0.0000   
16              0.0000        0.0000           0.000           0.0000   
17              0.0000        0.0000           0.000           0.0000   
18              0.0000        0.0000           0.000           0.0000   
19              0.0000        0.0000           0.000           0.0000   
20              0.0000        0.0000           0.000           0.0000   
21              0.0000        0.0000           0.000           0.0000   
22              0.0000        0.0000           0.000           0.0000   
23              0.0000        0.0000           0.000           0.0000   
24              0.0000        0.0000           0.000           0.0000   
25              0.0000        0.0000           0.000           0.0000   
26              0.0000        0.0000           0.000           0.0000   
27             

In [145]:
num_top_venues = 5

for hood in Toronto_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
                 venue  freq
0          Coffee Shop  0.08
1                 Café  0.06
2           Steakhouse  0.06
3  American Restaurant  0.06
4     Asian Restaurant  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.08
1        Cocktail Bar  0.06
2            Beer Bar  0.04
3  Seafood Restaurant  0.04
4         Cheese Shop  0.04


----Brockton, Exhibition Place, Parkdale Village----
                    venue  freq
0          Breakfast Spot  0.09
1                    Café  0.09
2             Coffee Shop  0.09
3  Furniture / Home Store  0.04
4           Burrito Place  0.04


----Business Reply Mail Processing Centre 969 Eastern----
                  venue  freq
0    Light Rail Station  0.11
1           Yoga Studio  0.06
2            Restaurant  0.06
3  Fast Food Restaurant  0.06
4            Smoke Shop  0.06


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----

From the above result, if we would like to enter a neighbour which has less competition, we should open our shop in:
1, Business Reply Mail Processing Centre 969 Eastern
2, Chinatown, Grange Park, Kensington Market
3, Christie
4,Dovercourt Village, Dufferin
5, Harbord, University of Toronto